python_modules/dagster/dagster/core/definitions/unresolved_asset_job_definition.py

```diff
@@ -0,0 +1,111 @@
+from typing import TYPE_CHECKING, Any, Dict, NamedTuple, Optional, Sequence, Union, cast
+
+import dagster._check as check
+from dagster.core.definitions.asset_layer import build_asset_selection_job
+from dagster.core.selector.subset_selector import parse_clause
+
+if TYPE_CHECKING:
+    from dagster.core.asset_defs import AssetsDefinition, SourceAsset
+    from dagster.core.asset_defs.asset_selection import AssetSelection
+    from dagster.core.definitions import ExecutorDefinition, JobDefinition
+
+
+class UnresolvedAssetJobDefinition(
+    NamedTuple(
+        "_UnresolvedAssetJobDefinition",
+        [
+            ("name", str),
+            ("selection", "AssetSelection"),
+            ("executor_def", Optional["ExecutorDefinition"]),
+            ("description", Optional[str]),
+            ("tags", Optional[Dict[str, Any]]),
+        ],
+    )
+):
+    def __new__(
+        cls,
+        name: str,
+        selection: "AssetSelection",
+        executor_def: Optional["ExecutorDefinition"] = None,
+        description: Optional[str] = None,
+        tags: Optional[Dict[str, Any]] = None,
+    ):
+        from dagster.core.asset_defs.asset_selection import AssetSelection
+        from dagster.core.definitions.executor_definition import ExecutorDefinition
+
+        return super(UnresolvedAssetJobDefinition, cls).__new__(
+            cls,
+            name=check.str_param(name, "name"),
+            selection=check.inst_param(selection, "selection", AssetSelection),
+            executor_def=check.opt_inst_param(executor_def, "executor_def", ExecutorDefinition),
+            description=check.opt_str_param(description, "description"),
+            tags=check.opt_dict_param(tags, "tags"),
+        )
+
+    def resolve(
+        self, assets: Sequence["AssetsDefinition"], source_assets: Sequence["SourceAsset"]
+    ) -> "JobDefinition":
+        """
+        Resolve this UnresolvedAssetJobDefinition into a JobDefinition.
+        """
+        return build_asset_selection_job(
+            name=self.name,
+            assets=assets,
+            source_assets=source_assets,
+            executor_def=self.executor_def,
+            description=self.description,
+            tags=self.tags,
+            asset_selection=self.selection.resolve(assets),
+        )
+
+
+def _selection_from_string(string: str) -> "AssetSelection":
+    from dagster.core.asset_defs.asset_selection import AssetSelection
+
+    if string == "*":
+        return AssetSelection.all()
+
+    parts = parse_clause(string)
+    if not parts:
+        check.failed(f"Invalid selection string: {string}")
+    u, item, d = parts
+
+    selection: AssetSelection = AssetSelection.keys(item)
+    if u:
+        selection = selection.upstream(u)
+    if d:
+        selection = selection.downstream(d)
+    return selection
+
+
+def define_asset_job(
+    name: str,
+    selection: Union[str, Sequence[str], "AssetSelection"] = None,
+    executor_def: "ExecutorDefinition" = None,
```

Comment:
Putting an executor on an asset job feels scary to me, because it means that executing the job inside vs. outside of the job has subtle differences that are difficult to detect ahead of time. I wonder if we should punt on this or mark it as experimental until we can do a little more diligence.